Race Test TestSubscribeEnvelopesByTopic is flaky

```
=== RUN   TestSubscribeEnvelopesByTopic
    store.go:52: creating database test_newtestapiserver_ivzyrrvagvzh ...
2026-04-01T07:13:04.608Z	INFO	registrant/registrant.go:59	registrant identified	{"originator_id": 100, "public_key": "0x02b4a6f3ba38981ff23a05d4d0a3710ebfadfc9b376e8a208ee280e92685e1641f"}
2026-04-01T07:13:04.778Z	INFO	api	api/server.go:115	creating api server
2026-04-01T07:13:04.778Z	INFO	publish-worker	message/publish_worker.go:54	starting
2026-04-01T07:13:04.778Z	INFO	publish-worker	message/publish_worker.go:116	started
2026-04-01T07:13:04.815Z	INFO	api	api/server.go:210	health handler registered
2026-04-01T07:13:04.815Z	INFO	api	api/server.go:222	reflection handler v1 registered
2026-04-01T07:13:04.815Z	INFO	api	api/server.go:228	reflection handler v1 alpha registered
2026-04-01T07:13:04.817Z	INFO	api	api/server.go:173	started api server	{"address": "127.0.0.1:43453"}
    subscribe_test.go:176: 
        	Error Trace:	/home/runner/work/xmtpd/xmtpd/pkg/api/message/subscribe_test.go:176
        	            				/home/runner/work/xmtpd/xmtpd/pkg/api/message/subscribe_test.go:222
        	Error:      	Should be true
        	Test:       	TestSubscribeEnvelopesByTopic
        	Messages:   	received unexpected update: nodeID=%!s(int32=200) seqID=1
2026-04-01T07:13:05.286Z	INFO	api	api/server.go:181	closing
2026-04-01T07:13:05.286Z	INFO	publish-worker	message/publish_worker.go:132	stopped
2026-04-01T07:13:05.286Z	INFO	api	api/server.go:195	closed
2026-04-01T07:13:05.312Z	ERROR	subscribe-worker	message/envelope_list.go:73	failed to get envelopes	{"error": "failed to connect to `user=postgres database=test_newtestapiserver_ivzyrrvagvzh`:\n\t127.0.0.1:8765 (localhost): failed SASL auth: context canceled\n\t[::1]:8765 (localhost): dial error: dial tcp [::1]:8765: operation was canceled", "originator_id": 11}
github.com/xmtp/xmtpd/pkg/api/message.(*subscriptionHandler).newSubscription.func2
	/home/runner/work/xmtpd/xmtpd/pkg/api/message/envelope_list.go:73
github.com/xmtp/xmtpd/pkg/db.(*DBSubscription[...]).poll
	/home/runner/work/xmtpd/xmtpd/pkg/db/subscription.go:102
github.com/xmtp/xmtpd/pkg/db.(*DBSubscription[...]).Start.func1
	/home/runner/work/xmtpd/xmtpd/pkg/db/subscription.go:81
--- FAIL: TestSubscribeEnvelopesByTopic (2.09s)
```

--- a/pkg/api/message/envelope_list.go
+++ b/pkg/api/message/envelope_list.go
@@ -13,11 +13,20 @@ import (
 )
 
 type subscriptionHandler struct {
-	mu         sync.Mutex
+	sync.Mutex
 	logger     *zap.Logger
 	store      *db.Handler
 	mergedSubs *funnel[[]queries.SelectGatewayEnvelopesBySingleOriginatorRow]
 	cursor     db.VectorClock
+	subs       map[uint32]*envelopePoller
+}
+
+type envelopePoller struct {
+	cancel context.CancelFunc
+	// TODO: Check - queries.GatewayEnvelopesView and queries.SelectGatewayEnvelopesByOriginatorsRow
+	// models are identical and they're overly verbose.
+	ch  <-chan []queries.SelectGatewayEnvelopesBySingleOriginatorRow
+	sub *db.DBSubscription[queries.SelectGatewayEnvelopesBySingleOriginatorRow, int64]
 }
 
 func newSubscriptionHandler(
@@ -30,6 +39,7 @@ func newSubscriptionHandler(
 		store:      store,
 		cursor:     cursor,
 		mergedSubs: newFunnel[[]queries.SelectGatewayEnvelopesBySingleOriginatorRow](),
+		subs:       make(map[uint32]*envelopePoller),
 	}
 }
 
@@ -99,10 +109,18 @@ func (s *subscriptionHandler) newSubscription(ctx context.Context, id uint32) (r
 		return fmt.Errorf("could not start subscription (id: %v): %w", id, err)
 	}
 
-	// Save the poller in the subscription handler.
-	s.mu.Lock()
-	defer s.mu.Unlock()
+	// Per node/originator poller and cancellation.
+	e := &envelopePoller{
+		cancel: cancel,
+		ch:     ch,
+		sub:    sub,
+	}
 
+	// Save the poller in the subscription handler.
+	s.Lock()
+	defer s.Unlock()
+
+	s.subs[id] = e
 	s.mergedSubs.addChannel(ch)
 
 	return nil
@@ -110,7 +128,7 @@ func (s *subscriptionHandler) newSubscription(ctx context.Context, id uint32) (r
 
 // allSubscriptions returns a channel merging all individual subscription channels.
 func (s *subscriptionHandler) allSubscriptions() <-chan []queries.SelectGatewayEnvelopesBySingleOriginatorRow {
-	s.mu.Lock()
-	defer s.mu.Unlock()
+	s.Lock()
+	defer s.Unlock()
 	return s.mergedSubs.output()
 }

--- a/pkg/api/message/export_test.go
+++ b/pkg/api/message/export_test.go
@@ -1,0 +1,39 @@
+package message
+
+import (
+	"context"
+	"time"
+
+	"github.com/xmtp/xmtpd/pkg/db"
+)
+
+// AwaitCursor blocks until the subscribe worker has polled past all sequence IDs in vc.
+// Only compiled during testing (export_test.go pattern).
+func (s *Service) AwaitCursor(ctx context.Context, vc db.VectorClock) error {
+	const checkInterval = 5 * time.Millisecond
+	ticker := time.NewTicker(checkInterval)
+	defer ticker.Stop()
+
+	for {
+		if s.subscribeWorker.subscriptions.cursorMet(vc) {
+			return nil
+		}
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ticker.C:
+		}
+	}
+}
+
+func (s *subscriptionHandler) cursorMet(vc db.VectorClock) bool {
+	s.Lock()
+	defer s.Unlock()
+	for nodeID, minSeq := range vc {
+		poller, ok := s.subs[nodeID]
+		if !ok || uint64(poller.sub.LastSeen()) < minSeq {
+			return false
+		}
+	}
+	return true
+}

--- a/pkg/api/message/subscribe_test.go
+++ b/pkg/api/message/subscribe_test.go
@@ -21,7 +21,6 @@ import (
 	"github.com/xmtp/xmtpd/pkg/migrator"
 	"github.com/xmtp/xmtpd/pkg/proto/xmtpv4/envelopes"
 	"github.com/xmtp/xmtpd/pkg/proto/xmtpv4/message_api"
-	message_apiconnect "github.com/xmtp/xmtpd/pkg/proto/xmtpv4/message_api/message_apiconnect"
 	"github.com/xmtp/xmtpd/pkg/registry"
 	"github.com/xmtp/xmtpd/pkg/testutils"
 	testUtilsApi "github.com/xmtp/xmtpd/pkg/testutils/api"
@@ -37,9 +36,7 @@ var (
 	allRows = make([]queries.InsertGatewayEnvelopeParams, 0)
 )
 
-func setupTest(
-	t *testing.T,
-) (message_apiconnect.ReplicationApiClient, *sql.DB, testUtilsApi.APIServerMocks) {
+func setupTest(t *testing.T) *testUtilsApi.APIServerTestSuite {
 	var (
 		nodes = []registry.Node{
 			{NodeID: 100, IsCanonical: true},
@@ -104,14 +101,19 @@ func setupTest(
 		},
 	}
 
-	return suite.ClientReplication, suite.DB, suite.APIServerMocks
+	return suite
 }
 
-func insertInitialRows(t *testing.T, store *sql.DB) {
-	testutils.InsertGatewayEnvelopes(t, store, []queries.InsertGatewayEnvelopeParams{
+func insertInitialRows(t *testing.T, suite *testUtilsApi.APIServerTestSuite) {
+	t.Helper()
+	testutils.InsertGatewayEnvelopes(t, suite.DB, []queries.InsertGatewayEnvelopeParams{
 		allRows[0], allRows[1],
 	})
-	time.Sleep(message.SubscribeWorkerPollTime + 100*time.Millisecond)
+	// Wait until the subscribe worker has polled past the inserted rows so that
+	// a subsequent subscription with LastSeen=nil won't see them.
+	ctx, cancel := context.WithTimeout(t.Context(), 500*time.Millisecond)
+	defer cancel()
+	require.NoError(t, suite.MessageService.AwaitCursor(ctx, db.VectorClock{100: 1, 200: 1}))
 }
 
 func insertAdditionalRows(t *testing.T, store *sql.DB, notifyChan ...chan bool) {
@@ -203,11 +205,11 @@ func validateUpdates(
 }
 
 func TestSubscribeEnvelopesByTopic(t *testing.T) {
-	client, store, _ := setupTest(t)
-	insertInitialRows(t, store)
+	suite := setupTest(t)
+	insertInitialRows(t, suite)
 
 	ctx := t.Context()
-	stream, err := client.SubscribeEnvelopes(
+	stream, err := suite.ClientReplication.SubscribeEnvelopes(
 		ctx,
 		connect.NewRequest(&message_api.SubscribeEnvelopesRequest{
 			Query: &message_api.EnvelopesQuery{
@@ -218,16 +220,16 @@ func TestSubscribeEnvelopesByTopic(t *testing.T) {
 	)
 	require.NoError(t, err)
 
-	insertAdditionalRows(t, store)
+	insertAdditionalRows(t, suite.DB)
 	validateUpdates(t, stream, []int{4})
 }
 
 func TestSubscribeEnvelopesByOriginator(t *testing.T) {
-	client, db, _ := setupTest(t)
-	insertInitialRows(t, db)
+	suite := setupTest(t)
+	insertInitialRows(t, suite)
 
 	ctx := t.Context()
-	stream, err := client.SubscribeEnvelopes(
+	stream, err := suite.ClientReplication.SubscribeEnvelopes(
 		ctx,
 		connect.NewRequest(&message_api.SubscribeEnvelopesRequest{
 			Query: &message_api.EnvelopesQuery{
@@ -238,16 +240,16 @@ func TestSubscribeEnvelopesByOriginator(t *testing.T) {
 	)
 	require.NoError(t, err)
 
-	insertAdditionalRows(t, db)
+	insertAdditionalRows(t, suite.DB)
 	validateUpdates(t, stream, []int{2, 4})
 }
 
 func TestSimultaneousSubscriptions(t *testing.T) {
-	client, store, _ := setupTest(t)
-	insertInitialRows(t, store)
+	suite := setupTest(t)
+	insertInitialRows(t, suite)
 
 	ctx := t.Context()
-	stream1, err := client.SubscribeEnvelopes(
+	stream1, err := suite.ClientReplication.SubscribeEnvelopes(
 		ctx,
 		connect.NewRequest(&message_api.SubscribeEnvelopesRequest{
 			Query: &message_api.EnvelopesQuery{},
@@ -255,7 +257,7 @@ func TestSimultaneousSubscriptions(t *testing.T) {
 	)
 	require.NoError(t, err)
 
-	stream2, err := client.SubscribeEnvelopes(
+	stream2, err := suite.ClientReplication.SubscribeEnvelopes(
 		ctx,
 		connect.NewRequest(&message_api.SubscribeEnvelopesRequest{
 			Query: &message_api.EnvelopesQuery{
@@ -266,7 +268,7 @@ func TestSimultaneousSubscriptions(t *testing.T) {
 	)
 	require.NoError(t, err)
 
-	stream3, err := client.SubscribeEnvelopes(
+	stream3, err := suite.ClientReplication.SubscribeEnvelopes(
 		ctx,
 		connect.NewRequest(&message_api.SubscribeEnvelopesRequest{
 			Query: &message_api.EnvelopesQuery{
@@ -277,18 +279,18 @@ func TestSimultaneousSubscriptions(t *testing.T) {
 	)
 	require.NoError(t, err)
 
-	insertAdditionalRows(t, store)
+	insertAdditionalRows(t, suite.DB)
 	validateUpdates(t, stream1, []int{})
 	validateUpdates(t, stream2, []int{2, 3})
 	validateUpdates(t, stream3, []int{3})
 }
 
 func TestSubscribeEnvelopesFromCursor(t *testing.T) {
-	client, store, _ := setupTest(t)
-	insertInitialRows(t, store)
+	suite := setupTest(t)
+	insertInitialRows(t, suite)
 
 	ctx := t.Context()
-	stream, err := client.SubscribeEnvelopes(
+	stream, err := suite.ClientReplication.SubscribeEnvelopes(
 		ctx,
 		connect.NewRequest(&message_api.SubscribeEnvelopesRequest{
 			Query: &message_api.EnvelopesQuery{
@@ -299,16 +301,16 @@ func TestSubscribeEnvelopesFromCursor(t *testing.T) {
 	)
 	require.NoError(t, err)
 
-	insertAdditionalRows(t, store)
+	insertAdditionalRows(t, suite.DB)
 	validateUpdates(t, stream, []int{1, 4})
 }
 
 func TestSubscribeEnvelopesFromEmptyCursor(t *testing.T) {
-	client, store, _ := setupTest(t)
-	insertInitialRows(t, store)
+	suite := setupTest(t)
+	insertInitialRows(t, suite)
 
 	ctx := t.Context()
-	stream, err := client.SubscribeEnvelopes(
+	stream, err := suite.ClientReplication.SubscribeEnvelopes(
 		ctx,
 		connect.NewRequest(&message_api.SubscribeEnvelopesRequest{
 			Query: &message_api.EnvelopesQuery{
@@ -319,15 +321,15 @@ func TestSubscribeEnvelopesFromEmptyCursor(t *testing.T) {
 	)
 	require.NoError(t, err)
 
-	insertAdditionalRows(t, store)
+	insertAdditionalRows(t, suite.DB)
 	validateUpdates(t, stream, []int{0, 1, 4})
 }
 
 func TestSubscribeEnvelopesInvalidRequest(t *testing.T) {
 	var (
-		client, _, _ = setupTest(t)
-		ctx          = t.Context()
-		tests        = []struct {
+		suite = setupTest(t)
+		ctx   = t.Context()
+		tests = []struct {
 			name  string
 			query *message_api.EnvelopesQuery
 		}{
@@ -352,7 +354,7 @@ func TestSubscribeEnvelopesInvalidRequest(t *testing.T) {
 
 	for _, test := range tests {
 		t.Run(test.name, func(t *testing.T) {
-			stream, err := client.SubscribeEnvelopes(
+			stream, err := suite.ClientReplication.SubscribeEnvelopes(
 				ctx,
 				connect.NewRequest(&message_api.SubscribeEnvelopesRequest{
 					Query: test.query,

--- a/pkg/db/subscription.go
+++ b/pkg/db/subscription.go
@@ -3,6 +3,7 @@ package db
 import (
 	"context"
 	"errors"
+	"sync"
 	"time"
 
 	"github.com/xmtp/xmtpd/pkg/tracing"
@@ -30,10 +31,18 @@ type PollingOptions struct {
 type DBSubscription[ValueType any, CursorType any] struct {
 	ctx      context.Context
 	logger   *zap.Logger
+	mu       sync.RWMutex
 	lastSeen CursorType
 	options  PollingOptions
 	query    PollableDBQuery[ValueType, CursorType]
 	updates  chan<- []ValueType
+}
+
+// LastSeen returns the cursor position last successfully processed by the subscription.
+func (s *DBSubscription[ValueType, CursorType]) LastSeen() CursorType {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	return s.lastSeen
 }
 
 func NewDBSubscription[ValueType any, CursorType any](
@@ -127,7 +136,9 @@ func (s *DBSubscription[ValueType, CursorType]) poll(trigger string) {
 		}
 
 		totalResults += len(results)
+		s.mu.Lock()
 		s.lastSeen = lastID
+		s.mu.Unlock()
 		s.updates <- results
 
 		// If we have less results than allowed, it means there's currently no more items to retrieve.

--- a/pkg/testutils/api/api.go
+++ b/pkg/testutils/api/api.go
@@ -120,6 +120,7 @@ type APIServerTestSuite struct {
 	ClientMetadata    metadata_apiconnect.MetadataApiClient
 	DB                *sql.DB
 	APIServerMocks    APIServerMocks
+	MessageService    *message.Service
 }
 
 // APIServerTestConfig allows explicitly setting some components used for tests.
@@ -207,10 +208,11 @@ func NewTestAPIServer(
 
 	authInterceptor := server.NewServerAuthInterceptor(jwtVerifier, log)
 
+	var replicationService *message.Service
 	serviceRegistrationFunc := func(mux *http.ServeMux, interceptors ...connect.Interceptor) (servicePaths []string, err error) {
 		interceptors = append(interceptors, authInterceptor)
 
-		replicationService, err := message.NewReplicationAPIService(
+		replicationService, err = message.NewReplicationAPIService(
 			ctx,
 			log,
 			registrant,
@@ -315,5 +317,6 @@ func NewTestAPIServer(
 		ClientPayer:       clientPayer,
 		ClientMetadata:    clientMetadata,
 		DB:                db.DB(),
+		MessageService:    replicationService,
 	}
 }

